### DISCUSSION TITLE
create and catch custom signals

### DISCUSSION THREAD

**Discussion by wedgebo** [pybricks/support#1431] [Q&A]
Using spike 3 I can create a custom signal and use this signal to trigger a different block.

I was wondering  if there is a similar mechanism in pybricks. Coding with blocks has triggered me, but I am still unsure if moving from spike3 to pybricks would cause more pain than joy.

I hope that I am too blind (or dumb) to realize how it is done. But, well in both cases can someone please elaborate on that topic using blocks or python code?

**Comment by laurensvalk** [comment]
Can you describe the end goal for what you'd like to code?

Sometimes, signals are used as a somewhat complicated means to do two things at the same time (because Spike 3 doesn't have easier ways of doing it.)  To cover such use cases, we've added the multitask block, which makes such things a lot easier and flexible.

Depending on your use case, perhaps we can make an example to illustrate how you could use it.

Or likewise, if something can't be done with this block, we can consider adding new types of blocks :slightly_smiling_face:

**Comment by wedgebo** [comment]
ok I will try to describe what I had in mind. There is no actual problem involved.

Given there is a block that computes something. Maybe even asynchronoulsy. If that result would cause different actions I could 'switch' the result or I could 'if.. else' it.
A cleaner solution could be to write tiny blocks for each possible result that needs to be treated and have a 'run on that specific signal' as a trigger.

**Reply by laurensvalk** [reply]
I think this can be done since you can make multiple starting points that wait on conditions, but we may be able to make this simpler with fewer blocks. Generally, if there are convincing use cases, we will definitely consider adding blocks.

So when you get to such a coding problem, we'll certainly want to have a look.

Here's a very different use case --- consider this thought experiment. Suppose you have mission A (driving in a square), mission B (such as operating a gripper mechanism), and mission C (like playing a sound), and suppose you want to use them as follows:

```
repeat 4 times:
     B
     A & B (at the same time)
     C
```

This is relatively hard to achieve with signals or other tools available in Scratch, but it is very easy to do with the Pybricks multitask block. The same is true for stopping tasks midway.

**Comment by wedgebo** [comment]
> I think this can be done since you can make multiple starting points that wait on conditions

Will take some time to figure out what direction that is going.  I think I didn't get the point yet.
Maybe I will need to get a single user license of blocks to have a solid platform for discussions and verification.

**Reply by laurensvalk** [reply]
Feel free to use the trial version via Patreon if you prefer. Thanks for considering supporting the project.

Happy to hear your thoughts! Since we can also add functionality to the underlying Python code as needed for new blocks, things are pretty flexible. We do try to strike the right balance between features and complexity, but we're definitely interested in making things work even better for robotics teams too.

**Comment by wedgebo** [comment]
Would the whole thing I described (or what the signal mechanism in spike3 does) comply to structuring a block program into functions?
I mean afaik calling functions is not part of blocks, so sending a signal and triggering a block with that signal reduces to simply calling a block i.e. function. Right?

However the complexity for implementation when it comes to normal or async function is beyond my scope....

**Reply by laurensvalk** [reply]
You can easily make and call functions, but it isn't entirely clear to me what you would like to do. A concrete example would probably help. :)

**Comment by wedgebo** [comment]
Apologies and thanks for your patience!

I really can't give a concrete example just because there is none. I am just imagining what could be done and how it could be done. That is wasting time.
my roadmap:
-get a blocks license :-)
-get used to block programming
-get used to the python api
-find the time to play with everything
-figure out an example, how stupid it might be
-ask for advice

I see no other option. However I will keep this thread open just in case.

P.S. I can spoil that chances are very high that I will post and ask for a feature request.... Serial port would be more than helpful for integration other microcontroller boards I worked and in the past and use lego hubs as imu/motor driver only.

**Reply by laurensvalk** [reply]
> Serial port would be more than helpful

If you mean UART via the sensor ports, see https://github.com/pybricks/support/issues/220

If you mean USB support, see https://github.com/pybricks/pybricks-micropython/pull/208

> integration other microcontroller boards I worked and in the past and use lego hubs as imu/motor driver only.

A good existing option for this might be via Bluetooth. Hub to hub communication [works like this](https://www.youtube.com/watch?v=WzmcihSV2YE) and any device like an Arduino or Raspberry Pi could participate in the network if they use the same [scanning/advertising protocol](https://github.com/pybricks/technical-info).

> -get a blocks license :-)
> -get used to block programming
> -get used to the python api
> -find the time to play with everything
> -figure out an example, how stupid it might be
> -ask for advice

Thanks!
